Create a ref-assembly package for HttpHandler

pranavkm opened the pull request:
Fixes https://github.com/dotnet/aspnetcore/issues/19202

------------------------

This adds a package to carry a ref assembly for WebAssemblyHttpHandler. The current contents resemble what's currently in the runtime, we can update it once the HttpHandler changes are in the runtime. I choose `Microsoft.AspNetCore.Components.WebAssembly.HttpHandler` as the package name since it's meant to be a stop-gap until the runtime package is available in .NET 5.0. We can move this code over to dotnet/Blazor once we're ok with the shape of this change.

pranavkm committed:
Create a ref-assembly package for HttpHandler

pranavkm committed:
Tweak versions?

pranavkm committed:
WIP

pranavkm committed:
Fixup

pranavkm committed:
Add APIs to configure browser fetch options

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Http/BrowserRequestCache.cs:
```diff
@@ -0,0 +1,44 @@
+        Reload,
+
+        /// <summary>
+        /// The browser looks for a matching request in its HTTP cache.
```
Looks like a wrongly copy/pasted description.

pranavkm replied:
I've copied the description verbatim from https://developer.mozilla.org/en-US/docs/Web/API/Request/cache

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.nuspec:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I should delete this. We rely on the project file to generate this.

pranavkm commented on src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+                throw new ArgumentNullException(nameof(requestMessage));
+            }
+
+            requestMessage.Properties["WebAssemblyReadResponseAsStream"] = true;
```
This string needs to be updated: `WebAssemblyEnableStreamingResponse`

SteveSandersonMS replied:
```suggestion
            requestMessage.Properties["WebAssemblyReadResponseAsStream"] = enabled;
```

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+        /// <remarks>
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/Request/integrity
+        /// </remarks>
+        public static HttpRequestMessage SetBrowserIntegrity(this HttpRequestMessage requestMessage, string integrity)
```
```suggestion
        public static HttpRequestMessage SetBrowserRequestIntegrity(this HttpRequestMessage requestMessage, string integrity)
```

All the preceding methods have "Request" in the name

SteveSandersonMS replied:
Also helps to clarify it's not the integrity of the browser we're checking, but of this individual request

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+        /// <remarks>
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
+        /// </remarks>
+        public static HttpRequestMessage SetBrowserFetchOption(this HttpRequestMessage requestMessage, string name, object value)
```
```suggestion
        public static HttpRequestMessage SetBrowserRequestOption(this HttpRequestMessage requestMessage, string name, object value)
```

Similar reason to above.

Also, in JavaScript the interface defining the options is called `RequestInit`, so it sounds right to use the term "request" rather than "fetch" from a JS perspective too.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+        /// Configures a value for the HTTP request.
+        /// </summary>
+        /// <param name="requestMessage">The <see cref="HttpRequestMessage"/>.</param>
+        /// <param name="name">The name of the HTTP fetch option.</param>
```
```suggestion
        /// <param name="name">The name of the option, which should correspond to a key defined on https://fetch.spec.whatwg.org/#requestinit</param>
```

Not sure if URLs in XML docs can be made clickable through any special syntax.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:
```diff
@@ -0,0 +1,169 @@
+        /// This API is only effective when the browser HTTP Fetch supports streaming.
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream.
+        /// </remarks>
+        public static HttpRequestMessage EnableStreamingResponse(this HttpRequestMessage requestMessage)
```
```suggestion
        public static HttpRequestMessage SetBrowserResponseStreamingEnabled(this HttpRequestMessage requestMessage, bool enabled)
```

Name suggestion is so it appears alphabetically with the others in intellisense.

Suggestion to make it possible to set "false" is because maybe in the future it might default to true, and you need a way to turn it off. Also with all the other methods here, it's possible to undo them by overwriting with `null` or similar, so it's nice if this one is undoable too.

What do you think?

SteveSandersonMS commented on src/Components/WebAssembly/testassets/StandaloneApp/Pages/FetchData.razor:
```diff
@@ -52,8 +52,13 @@ else
-        forecasts = await Http.GetFromJsonAsync<WeatherForecast[]>(
-            $"sample-data/weather.json?date={startDate.ToString("yyyy-MM-dd")}");
+        var requestMessage = new HttpRequestMessage(HttpMethod.Get, $"sample-data/weather.json?date={startDate.ToString("yyyy-MM-dd")}");
+        requestMessage.SetBrowserRequestCache(BrowserRequestCache.NoCache);
```
Please don't use `Components/WebAssembly/testassets/StandaloneApp` for testing or demonstrating random features :) It's meant to stay close to the default project template, as it's really a sandbox project not a test asset.

If you're adding this for an E2E test, could you do it inside `Components\test\testassets\BasicTestApp` instead?

SteveSandersonMS replied:
Also, there are a couple of redundant blank lines below.

SteveSandersonMS commented on src/Components/WebAssembly/testassets/StandaloneApp/_Imports.razor:
```diff
@@ -2,5 +2,6 @@
 @using System.Net.Http.Json
 @using Microsoft.AspNetCore.Components.Routing
 @using Microsoft.AspNetCore.Components.Web
+@using Microsoft.AspNetCore.Components.WebAssembly.Http
```
If this is also going into the project template then good, or if not then please revert.

SteveSandersonMS commented on src/Components/test/testassets/BasicTestApp/Program.cs:
```diff
@@ -28,16 +27,18 @@ public static async Task Main(string[] args)
                 // Needed because the test server runs on a different port than the client app,
                 // and we want to test sending/receiving cookies under this config
-                WebAssemblyHttpMessageHandlerOptions.DefaultCredentials = FetchCredentialsOption.Include;
+                httpClient = new HttpClient(new ConfigureForCorsHandler(new WebAssemblyHttpHandler()));
```
Now we have the capability of doing this on a per request basis, I would have a slight preference for moving the CORS-specific test logic into the CORS-specific test and not having it apply globally to the whole project.

Do you think that's feasible?

SteveSandersonMS commented:
BTW if the suggested API changes aren't possible because the refassembly has to correspond to an existing underlying implementation, totally fine to leave that until a future follow-up.

If so, and if you do agree that API changes are beneficial, would you be able to add an issue to track this?

pranavkm committed:
Apply suggestions from code review

Co-Authored-By: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/blazor build 20200410.3

- Microsoft.AspNetCore.Components.WebAssembly.Runtime: 3.2.0-preview5.20210.1 -> 3.2.0-preview5.20210.3

pranavkm committed:
Fixup

pranavkm committed:
Fixups

pranavkm committed:
Fixup doc comment

pranavkm commented on src/Components/test/testassets/BasicTestApp/HttpClientTest/CookieCounterComponent.razor:
```diff
@@ -1,3 +1,5 @@
+@using Microsoft.AspNetCore.Components.WebAssembly.Http
```
Do you think this import needs to be in `_Imports.razor`? Another alternative might be to move the extensions to `Microsoft.AspNetCore.Components.WebAssembly` instead, so it's more discoverable.

SteveSandersonMS replied:
Yeah, I'd definitely be fine with it going into `_Imports.razor` in the template. Then it doesn't have to occupy an especially prevalent namespace.

javiercn commented on eng/ProjectReferences.props:
```diff
@@ -12,6 +12,7 @@
     <ProjectReferenceProvider Include="Microsoft.AspNetCore.Components.DataAnnotations.Validation" ProjectPath="$(RepoRoot)src\Components\WebAssembly\Validation\src\Microsoft.AspNetCore.Components.DataAnnotations.Validation.csproj"  />
     <ProjectReferenceProvider Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" ProjectPath="$(RepoRoot)src\Components\WebAssembly\WebAssembly.Authentication\src\Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj"  />
     <ProjectReferenceProvider Include="Wasm.Performance.Driver" ProjectPath="$(RepoRoot)src\Components\benchmarkapps\Wasm.Performance\Driver\Wasm.Performance.Driver.csproj" RefProjectPath="$(RepoRoot)src\Components\benchmarkapps\Wasm.Performance\ref\Wasm.Performance.Driver.csproj" />
+    <ProjectReferenceProvider Include="Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler" ProjectPath="$(RepoRoot)src\Components\WebAssembly\WebAssemblyHttpHandler\src\Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj" RefProjectPath="$(RepoRoot)src\Components\WebAssembly\WebAssemblyHttpHandler\ref\Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj" />
```
typo

javiercn reviewed (commented):
Looks great!

I don't have a lot of context on this area, so I'll let Steve have the final say!

dougbu commented:
Nothing should have a ref/ assembly unless it's in the shared framework. We removed extra ref/ projects entirely in at least 'release/3.1' though that cleanup may not be complete in 'master' or 'blazor-wasm'. Why is a ref/ assembly necessary / wanted here?

Note ref/ assemblies only add baggage to shipped packages and a slight complication to assembly resolution on user's machines. The only reason would be hiding `public` API from user's compilers. But, we shouldn't need such complications.

Bottom line, add the src/ project but not the ref/ one.

SteveSandersonMS commented on src/Components/WebAssembly/WebAssemblyHttpHandler/src/Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj:
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Typo in the filename

SteveSandersonMS replied:
And various related filenames below

SteveSandersonMS reviewed (approved):
Excellent!

Would be great to fix the filename typos if possible

pranavkm committed:
Fixup typo

pranavkm committed:
Fixup

pranavkm commented:
Thanks @dougbu. I realized that the project isn't a reference provider due to how it builds. Updated it and undid the changes to `ProjectReferences.props`

pranavkm committed:
Add using to template

dougbu commented:
@pranavkm for future reference, there's a difference between being a reference provider (which means `@(Reference>` items in other projects work) and providing a ref/ **assembly** (which is about shipping something to customers in our targeting pack). And, you don't need long paths like `..\..\WebAssemblyHttpHandler\src\Microsoft.AspNetCore.Components.WebAssembly.HttpHandler.csproj` in multiple places when using `@(Reference)` items.
